### DISCUSSION TITLE
[FEATURE REQUEST] replace masking with inpainting (model weights download provided)

### DISCUSSION THREAD

**Discussion by ghost** [hlky/stable-diffusion-webui#213] [Feature suggestions]
download here: https://heibox.uni-heidelberg.de/f/4d9ac7ea40c64582b7c9/?dl=1

**Is your feature request related to a problem? Please describe.**
current masking feature doesn't use visual context so it looks out of place

**Describe the solution you'd like**
use inpainting with the official inpainting weights

**Describe alternatives you've considered**
not having inpainting

**Additional context**
taken from compvis repo for inpainting line:
`wget -O models/ldm/inpainting_big/last.ckpt https://heibox.uni-heidelberg.de/f/4d9ac7ea40c64582b7c9/?dl=1`

**Comment by hlky** [comment]
#179 

You can load a different checkpoint now with the config option, does this solve your issue?

**Comment by hlky** [comment]
Also masking itself is unrelated to inpainting.

I believe you meant to say img2img, which is also not quite the same as inpainting.

**Comment by ghost** [comment]
> Also masking itself is unrelated to inpainting.
> 
> I believe you meant to say img2img, which is also not quite the same as inpainting.

right, I'm requesting adding inpainting as a feature.  
inpainting could replace masking, as it's probably what most people want masking to do.

I've provided a download to the inpainting model weights according to the compvis repo (so it will be possible to test the feature)

**Comment by hlky** [comment]
Inpainting is an advanced feature and afaik it can't be implemented yet, I believe the model you are referring to is not a stable diffusion model, you may have found it in the code, but it's not release by stable diffusion.

If you have more information about implementation you should provide it.

**Comment by codedealer** [comment]
> download here: https://heibox.uni-heidelberg.de/f/4d9ac7ea40c64582b7c9/?dl=1

These are the old inpainting weights from latent diffusion, doubt they're going to be of much use in SD not without the serious work to be done.

**Reply by guwa1** [reply]
Isn't inpainting possible using stable diffusion via diffusers (https://github.com/huggingface/diffusers/tree/main/examples/inference) and thus also possible here? Or am I missing something?

**Reply by ghost** [reply]
> Isn't inpainting possible using stable diffusion via diffusers (https://github.com/huggingface/diffusers/tree/main/examples/inference) and thus also possible here? Or am I missing something?

yeah, it would appear @codedealer is correct (I was misinformed I apologize), but also what @guwa1 said.  

I'm not sure how useful the inpainting.py is since it relies on the diffusers library which seems to have the huggingface token requirement, which I'm not a big fan of.

**Comment by Sanster** [comment]
You can try [Lama Cleaner](https://github.com/Sanster/lama-cleaner), it integrates multiple inpainting models, including inpainting model from  [latent-diffusion](https://github.com/CompVis/latent-diffusion). 

![image](https://user-images.githubusercontent.com/3998421/186935390-30fef1a3-37c9-44b9-af6a-6fec1939f7a7.png)

**Reply by JoshuaKimsey** [reply]
Oh nice, it's like what my Google Pixel 6 Pro can do with the Magic Eraser feature! I actually really like that.

On this point though, I know I heard that in-painting was coming to DreamStudio, and maybe Stable Diffusion as a whole, at some point. Here's just some sources I found referencing this as being a thing, Emmad in particular is heavily suggesting it will be:

* https://twitter.com/genekogan/status/1555184488606564353?s=20&t=d2coxh_tJDvEVnE__Rv3VQ
* https://twitter.com/EMostaque/status/1561778925906395140?s=20&t=yIR-Bcq55OmWyksKhHLW_g
* https://twitter.com/EMostaque/status/1563348451060232192?s=20&t=yIR-Bcq55OmWyksKhHLW_g

These were just a couple I saw, but I've seen references to this elsewhere too. It'll be interesting to see it applied natively to Stable Diffusion once it's released.

**Comment by Muskelmagier** [comment]
https://github.com/basujindal/stable-diffusion has found a way to create an inpainting script could we ask them?

**Comment by telemole** [comment]
I've been using the mask feature here as crude as it might be, then going to photoshop with the original image and putting the newly created on a layer and painting out the unformed parts - decent workaround - would be great if I didn't have to though ;)

**Comment by Sanster** [comment]
[diffusers](https://github.com/huggingface/diffusers) has an example script for inpainting https://github.com/huggingface/diffusers/tree/main/examples/inference#in-painting-using-stable-diffusion

**Comment by fabwa** [comment]
https://github.com/basujindal/stable-diffusion/blob/main/scripts/inpaint.py

**Reply by telemole** [reply]
can anyone describe how to implement this? If I simply drop it in my scripts folder i'm not sure how to call it, let alone from the UI - would love to though! :) Please give some tips?

**Reply by codedealer** [reply]
needs trained model

**Reply by telemole** [reply]
so which model was the script noted above using? Apologies I'm new but trying to learn fast.  Is the script noted above not based on the same stable diffusion model we are using here? thanks for the help

**Reply by codedealer** [reply]
it's a placeholder by stability devs, the model is private for now

**Reply by telemole** [reply]
thank you! So there is a separate model called in the inpainting script that is not part of the stablediffusion 1.4 release that this UI is based on? I think I'm tracking now ;) WILL they be releasing it?

**Reply by codedealer** [reply]
I don't know but I don't see why won't they at some point in the future

**Comment by Sanster** [comment]
Finally make SD inpainting works in [lama-cleaner](https://github.com/Sanster/lama-cleaner). 

![image](https://user-images.githubusercontent.com/3998421/191783496-3de95ef1-f4b8-4afa-a61a-cc2bf8468a62.png)


To get a good results, you need right prompt and a lot of luck(maybe SD1.5 will do better job). Here are some of my test results

### Original Image
![overture-creations-5sI6fQgYIuo](https://user-images.githubusercontent.com/3998421/191783867-93fae1a5-8641-4637-a98c-c56d7cf653e1.png)

### Inpaint Result

![astronaut](https://user-images.githubusercontent.com/3998421/191783907-071f9e9b-317a-435a-977e-9fe857ada3d2.png)

![cat](https://user-images.githubusercontent.com/3998421/191784009-06b34773-af24-4f7e-91a7-13949a8bffeb.png)

![fox](https://user-images.githubusercontent.com/3998421/191784024-1fb79cf6-3bf7-4f1b-a3dc-623b3fd71ea1.png)

![ironman](https://user-images.githubusercontent.com/3998421/191784179-e8acfd93-93af-43c0-a584-5a84697b54c2.png)
